backend/cn/lib/eval.ml

```diff
@@ -0,0 +1,606 @@
+  if Stdlib.( = ) sgn1 sgn2 then
+    Result.error ("Mismatched signs at " ^ Locations.simple_location here)
+  else if sz1 <> sz2 then
+    Result.error ("Mismatched sizes at " ^ Locations.simple_location here)
+  else
+    Ok ()
+
+
+let handle_bounds (sgn, sz) (n : Z.t) : (Z.t, string) result =
+  let min, max = BT.bits_range (sgn, sz) in
+  match sgn with
+  | Unsigned ->
+    Ok
+      (if Z.gt n max then
+         Z.rem n (Z.add max Z.one)
+       else if Z.lt n Z.zero then
+         Z.add n max
+       else
+         n)
+  | Signed ->
+    if Z.lt n min then
+      Error "signed underflow"
+    else if Z.gt n max then
+      Error "signed overflow"
+    else
+      Ok n
+
+
+let eval_num_binop
+  (eval : IT.t -> (IT.t, string) result)
+  (bt : BT.t)
+  (here : Locations.t)
+  (op : IT.binop)
+  (f : Z.t -> Z.t -> Z.t)
+  (it1 : IT.t)
+  (it2 : IT.t)
+  : (IT.t, string) result
+  =
+  let ( let@ ) = Result.bind in
+  let return = Result.ok in
+  let@ it1 = eval it1 in
+  let@ it2 = eval it2 in
+  match (it1, it2) with
+  | IT (Const (Z n1), _, _), IT (Const (Z n2), _, _) ->
+    return @@ IT.num_lit_ (f n1 n2) bt here
+  | IT (Const (Bits ((sgn1, sz1), n1)), _, _), IT (Const (Bits ((sgn2, sz2), n2)), _, _)
+    ->
+    let@ () = check_bits_bt (sgn1, sz1) (sgn2, sz2) here in
+    (match sgn1 with
+     | Unsigned ->
+       let@ res = handle_bounds (sgn1, sz1) (f n1 n2) in
+       return @@ IT.num_lit_ res bt here
+     | Signed ->
+       let@ res = handle_bounds (sgn1, sz1) (f n1 n2) in
+       return @@ IT.num_lit_ res bt here)
+  | _ -> return @@ IT.arith_binop op (it1, it2) here
+
+
+let eval_term_generic
+  (eval_aux : IT.t -> (IT.t, string) result)
+  (prog5 : unit Mucore.mu_file)
+  (it : IT.t)
+  : (IT.t, string) result
+  =
+  let ( let@ ) = Result.bind in
+  let return = Result.ok in
+  let open IT in
+  let (IT (t_, bt, here)) = it in
+  let eval_num_binop = eval_num_binop eval_aux bt here in
+  match t_ with
+  | Const _ -> return it
+  | Sym x -> Result.error (Sym.pp_string x ^ " is free")
+  (* Unary ops *)
+  | Unop (Not, it') ->
+    let@ it' = eval_aux it' in
+    (match it' with
+     | IT (Const (Bool b), _, _) -> return @@ bool_ (not b) here
+     | _ -> return @@ not_ it' here)
+  | Unop (Negate, it') ->
+    let@ it' = eval_aux it' in
+    (match it' with
+     | IT (Const (Z n), _, _) -> return @@ num_lit_ (Z.neg n) bt here
+     | IT (Const (Bits ((Signed, sz), n)), _, _) ->
+       let@ res = handle_bounds (Signed, sz) (Z.neg n) in
+       return @@ num_lit_ res bt here
+     | IT (Const (Bits ((Unsigned, _), _)), _, _) ->
+       Result.error ("Can't negate unsigned integer at " ^ Locations.simple_location here)
+     | _ -> return @@ negate it' here)
+  | Unop (BW_CLZ_NoSMT, _) -> failwith "todo: BW_CLZ_NoSMT"
+  | Unop (BW_CTZ_NoSMT, it') ->
+    let@ it' = eval_aux it' in
+    (match it' with
+     | IT (Const (Bits ((_, sz), n)), _, _) ->
+       let res = Z.trailing_zeros n in
```

Comment:
Have to make sure Z's `trailing_zeros` agrees with how we've defined the SMT solver mapping in `solver.ml`.